File: app/src/main/java/com/example/myanimedata/ui/AllMangaFragment.java
package com.example.myanimedata.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.myanimedata.R;
import com.example.myanimedata.adapter.MangaAdapter;
import com.example.myanimedata.api.ApiClient;
import com.example.myanimedata.api.ApiService;
import com.example.myanimedata.api.MangaResponse;
import com.example.myanimedata.api.MangaResult;
import com.example.myanimedata.databinding.FragmentAllMangaBinding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class AllMangaFragment extends Fragment {
    private ApiService apiService;
    private FragmentAllMangaBinding binding;
    private MangaAdapter allMangaAdapter;

    private final List<MangaResult> allMangaResults = new ArrayList<>();

    private int page = 1;

    public AllMangaFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentAllMangaBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        Retrofit retrofit = ApiClient.getClient();
        apiService = retrofit.create(ApiService.class);

        GridLayoutManager mLayoutManager = new GridLayoutManager(getContext(), 3);

        RecyclerView rvAllMangaList = root.findViewById(R.id.rvAllMangaList);
        allMangaAdapter = new MangaAdapter(allMangaResults, getContext());

        rvAllMangaList.setLayoutManager(mLayoutManager);
        rvAllMangaList.setAdapter(allMangaAdapter);
        getAllMangaData(page);

        rvAllMangaList.addOnScrollListener(new RecyclerView.OnScrollListener(){
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy){
                super.onScrolled(recyclerView, dx, dy);

                if(!recyclerView.canScrollVertically(1)){
                    page++;
                    getAllMangaData(page);
                }
            }
        });

        return root;
    }

    private void getAllMangaData(int page) {
        int limit = 1;

        Call<MangaResponse> call = apiService.getAllManga(page, limit);
        call.enqueue(new Callback<MangaResponse>() {
            @Override
            public void onResponse(@NonNull Call<MangaResponse> call, @NonNull Response<MangaResponse> response) {
                if(response.body() != null){
                    int oldCount = allMangaResults.size();
                    binding.loadingAllManga.setVisibility(View.GONE);
                    binding.rvAllMangaList.setVisibility(View.VISIBLE);

                    allMangaResults.addAll(response.body().getMangaResults());
                    allMangaAdapter.notifyItemRangeInserted(oldCount, allMangaResults.size());
                } else {
                    binding.loadingAllManga.setVisibility(View.GONE);
                    binding.textNoAllResult.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<MangaResponse> call, @NonNull Throwable t) {
                binding.loadingAllManga.setVisibility(View.GONE);
                Toast.makeText(getContext(), "Fail to Fetch Data !!!",
                        Toast.LENGTH_SHORT).show();
            }
        });
    }


}

File: app/src/main/java/com/example/myanimedata/ui/AllMangaFragment.kt
package com.example.myanimedata.ui

import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myanimedata.R
import com.example.myanimedata.adapter.MangaAdapter
import com.example.myanimedata.api.ApiClient.client
import com.example.myanimedata.api.ApiService
import com.example.myanimedata.api.MangaResponse
import com.example.myanimedata.api.MangaResult
import com.example.myanimedata.databinding.FragmentAllMangaBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class AllMangaFragment : Fragment() {
    private var apiService: ApiService? = null
    private var binding: FragmentAllMangaBinding? = null

    private var allMangaAdapter: MangaAdapter? = null
    private val allMangaResults: MutableList<MangaResult> = ArrayList()

    private var page = 1

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        binding = FragmentAllMangaBinding.inflate(inflater, container, false)

        val root: View = binding!!.root
        val retrofit = client
        apiService = retrofit!!.create(ApiService::class.java)

        val mPortraitLayoutManager = GridLayoutManager(context, 3)
        val mLandscapeLayoutManager = GridLayoutManager(context, 5)
        val rvAllMangaList = root.findViewById<RecyclerView>(R.id.rvAllMangaList)
        allMangaAdapter = context?.let { MangaAdapter(allMangaResults, it) }

        if(resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
            rvAllMangaList.layoutManager = mLandscapeLayoutManager
        } else {
            rvAllMangaList.layoutManager = mPortraitLayoutManager
        }

        rvAllMangaList.adapter = allMangaAdapter
        rvAllMangaList.setHasFixedSize(true)
        getAllMangaData(page)

        rvAllMangaList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!recyclerView.canScrollVertically(1)) {
                    page++
                    getAllMangaData(page)
                }
            }
        })
        return root
    }

    private fun getAllMangaData(page: Int) {
        val limit = 15
        val call = apiService!!.getAllManga(page, limit)

        call!!.enqueue(object : Callback<MangaResponse?> {
            override fun onResponse(
                call: Call<MangaResponse?>,
                response: Response<MangaResponse?>
            ) {
                if (response.body() != null) {
                    val oldCount = allMangaResults.size
                    binding!!.loadingAllManga.visibility = View.GONE
                    binding!!.rvAllMangaList.visibility = View.VISIBLE
                    response.body()!!.mangaResults?.let { allMangaResults.addAll(it) }
                    allMangaAdapter!!.notifyItemRangeInserted(oldCount, allMangaResults.size)
                } else if (allMangaResults.isEmpty()) {
                    binding!!.loadingAllManga.visibility = View.GONE
                    binding!!.textNoAllResult.visibility = View.VISIBLE
                } else {
                    binding!!.loadingAllManga.visibility = View.GONE
                }
            }

            override fun onFailure(call: Call<MangaResponse?>, t: Throwable) {
                binding!!.loadingAllManga.visibility = View.GONE
                Toast.makeText(
                    context, t.message + " cause : "
                            + t.cause, Toast.LENGTH_SHORT
                ).show()
            }
        })
    }
}